Support NULLS FIRST/LAST options on ORDER BY

ggevay opened the pull request:
### Motivation

  * This PR adds a known-desirable feature: https://github.com/MaterializeInc/materialize/issues/12666

### Tips for reviewer

I separated the change into different commits for code, test, docs, generated doc files. (I will squash when merging the PR.)

### Testing

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Add options `NULLS LAST` and `NULLS FIRST` to `ORDER BY`.

philip-stoev commented:
I would like to give this a try, first thing tomorrow morning.

teskje commented on src/sql-parser/src/parser.rs:
```diff
@@ -4639,7 +4639,7 @@ impl<'a> Parser<'a> {
     }
 
-    /// Parse an expression, optionally followed by ASC or DESC (used in ORDER BY)
+    /// Parse an expression, optionally followed by ASC or DESC (used in ORDER BY), and then `[NULLS { FIRST | LAST }]`
```
Nit: I'd move the "(used in ORDER BY)" to the end, since it applies to both ASC/DESC and NULLS.

Also the line is a bit long for my taste. `rustfmt` doesn't muck with the line length of comments, so you need to take care of keeping that reasonable yourself. I think most people try to keep comments to 80 or 100 characters.

ggevay replied:
done

teskje commented on src/sql-parser/src/parser.rs:
```diff
@@ -4650,7 +4650,22 @@ impl<'a> Parser<'a> {
+                FIRST => false,
+                LAST => true,
+                _ => unreachable!(),
+            })
```
Nit: This could probably be simplified to:
```rust
let last = self.expect_one_of_keywords(&[FIRST, LAST])? == LAST;
Some(last)
```

ggevay replied:
done

teskje commented on src/sql/src/plan/query.rs:
```diff
@@ -3496,7 +3493,20 @@ where
+                desc,
+                nulls_last,
+            },
+        )
```
Nit: This might be a better fit for rustfmt's layouting:
```rust
.map(|(i, order)| ColumnOrder {
    column: i,
    desc: oder.desc,
    nulls_last: order.nulls_last,
})
```

... or with less boilerplate:
```rust
.map(|(i, order)| ColumnOrder {
    column: i,
    ..order
})
```

ggevay replied:
done

ggevay replied:
This is really nice, btw. :)

teskje commented on doc/user/content/sql/select.md:
```diff
@@ -69,7 +69,7 @@ _join&lowbar;expr_ | A join expression; for more details, see our [`JOIN` docume
 **OPTION (** _hint&lowbar;list_ **)** | Specify one or more [query hints](#query-hints).
 **HAVING** _expression_ | Filter aggregations by _expression_.
-**ORDER BY** _col&lowbar;ref_... | Order results in either **ASC** or **DESC** order (_**ASC** is implied default_).<br/><br>
+**ORDER BY** _col&lowbar;ref_... | Order results in either **ASC** or **DESC** order (_**ASC** is implied default_).<br/><br/>You can also specify whether to order _NULL_ values as last or as first. _(The default is last for **ASC**, and first for **DESC**.)_<br/><br>
```
I think that should be:
> You can also specify whether to order _NULL_ values ~~as~~ last or ~~as~~ first.

ggevay replied:
done

teskje reviewed (approved):
Looks great, except for a couple nits!

Let's wait for someone from the SQL team to have a look as well. I'm adding @sploiselle.

wangandi reviewed (commented):
We could ask somebody from the devex team for help reviewing the user doc changes.

I would prefer adding more `.slt` tests for lateral joins with an order by. The reason is that the ordering in a query like `SELECT * FROM (VALUES (1), (NULL), (3), (2)) ORDER BY column1 nulls first` is done by the `RowSetFinishing` struct that runs separately from the dataflow, so it shows up in a plan like this:
```
                      Optimized Plan                       
-----------------------------------------------------------
 %0 =                                                     +
 | Constant (null) (1) (2) (3)                            +
                                                          +
 Finish order_by=(#0 asc) limit=none offset=0 project=(#0)+
```
But the ordering in a lateral join is done by the `TopK` dataflow operator, so the plan looks like
```
query T multiline
EXPLAIN PLAN FOR SELECT state, name FROM
    (SELECT DISTINCT state FROM cities) grp,
    LATERAL (SELECT name, pop FROM cities WHERE state = grp.state ORDER BY pop DESC LIMIT 3)
----
Source materialize.public.cities (u1):
| Project (#0..=#2)
Query:
%0 =
| Get materialize.public.cities (u1)
| TopK group=(#1) order=(#2 desc nulls_first) limit=3 offset=0
| Project (#1, #0)

EOF
```

ggevay commented:
Thank you all for the comments! I addressed the comments, squashed the commits, and force pushed.

@wangandi 
I added one more test in `topk.slt` for `order by` inside a lateral join with explicit `NULLS FIRST`. So now there are two tests in `topk.slt` where it's checking query outputs for this case, and there are several more cases where it's just checking plans. Is this ok?

wangandi reviewed (approved):
Yes it should be sufficient to render in a dataflow both `nulls first` and `null last` and check that the output is correct.

ggevay commented:
@morsapaes , could you please take a look at the doc changes? (or suggest someone else) I changed a few SQL syntax diagrams.

morsapaes commented on doc/user/content/sql/select.md:
```diff
@@ -69,7 +69,7 @@ _join&lowbar;expr_ | A join expression; for more details, see our [`JOIN` docume
 **OPTION (** _hint&lowbar;list_ **)** | Specify one or more [query hints](#query-hints).
 **HAVING** _expression_ | Filter aggregations by _expression_.
-**ORDER BY** _col&lowbar;ref_... | Order results in either **ASC** or **DESC** order (_**ASC** is implied default_).<br/><br>
+**ORDER BY** _col&lowbar;ref_... | Order results in either **ASC** or **DESC** order (_**ASC** is implied default_).<br/><br/>You can also specify whether to order _NULL_ values last or first. _(The default is last for **ASC**, and first for **DESC**.)_<br/><br>
```
```suggestion
**ORDER BY** _col&lowbar;ref_... | Sort results in either **ASC** or **DESC** order (_default: **ASC**_).<br/><br/>Use the **NULLS FIRST** and **NULLS LAST** options to determine whether nulls appear before or after non-null values in the sort ordering _(default: **NULLS LAST** for **ASC**, **NULLS FIRST** for **DESC**).<br/><br>
```

ggevay replied:
Thx!

philip-stoev reviewed (approved):
Thank you. Seems to work as advertised. I tried this in the context of window functions as well, which is yet another place `ORDER BY` can appear:

```
WITH t (x) AS (VALUES ('a'), (NULL), ('b'), ('c'))
SELECT row_number() OVER (ORDER BY x NULLS FIRST), x FROM t;

WITH t (x) AS (VALUES ('a'), (NULL), ('b'), ('c'))
SELECT row_number() OVER (ORDER BY x NULLS LAST), x FROM t;



DROP TABLE IF EXISTS t;
CREATE TABLE t (x TEXT);
INSERT INTO t VALUES  ('a'), (NULL), ('b'), ('c');
SELECT row_number() OVER (ORDER BY x NULLS FIRST), x FROM t;
SELECT row_number() OVER (ORDER BY x NULLS LAST), x FROM t;
```

I also tried creating views and then using show create view to make sure that the nulls first / nulls last modifier is preserved in the view definition.

If you want to stick the queries above in some .slt file that would be much appreciated.

ggevay commented:
Thanks @philip-stoev, done!
